Base/src/main/java/io/deephaven/base/ringbuffer/AggregatingCharRingBuffer.java

```diff
@@ -0,0 +1,448 @@
+        final long intersectionSize = calcTail > head ? calcTail - head : 0;
+
+        // Now r1 and r2 are the two ranges that need to be recalculated.
+        // r1 needs to be recalculated because the values have been reset to identityVal.
+        // r2 needs to be recalculated because the values are new.
+        long r1Head = calcHead;
+        long r1Tail = calcTail - intersectionSize;
+        long r2Head = head + intersectionSize;
+        long r2Tail = tail;
+
+        final long newBase = r1Head; // aka calcHead
+
+        r1Head -= newBase; // aka 0
+        r1Tail -= newBase;
+        r2Head -= newBase; // >= 0 because head is always ahead of calcHead
+        r2Tail -= newBase;
+
+        final long r2Size = r2Tail - r2Head;
+        r2Head = r2Head & mask; // aka r2Head % capacity
+        r2Tail = r2Head + r2Size;
+
+        if (r2Tail <= storage.length) {
+            // R2 is a single segment in the "normal" direction
+            // with no wrapping. You're in one of these cases
+            // [----------) R1
+            // [--) case 1 (subsume)
+            // [----) case 2 (extend)
+            // [--) case 3 (two ranges)
+
+            if (r2Tail <= r1Tail) { // case 1: subsume
+                r2Head = r2Tail = 0; // empty
+            } else if (r2Head <= r1Tail) { // case 2: extend
+                r1Tail = r2Tail;
+                r2Head = r2Tail = 0; // empty
+            }
+            // else : R1 and R2 are correct as is.
+        } else {
+            // R2 crosses the modulus. We can think of it in two parts as
+            // part 1: [r2ModHead, capacity)
+            // part 2: [capacity, r2ModTail)
+            // If we shift it left by capacity (note: values become negative which is ok)
+            // then we have
+            // part 1: [r2ModHead - capacity, 0)
+            // part 2: [0, r2ModTail - capacity)
+
+            // Because it's now centered at 0, it is suitable for extending
+            // R1 on both sides. We do all this (adjust R2, extend R1, reverse the adjustment)
+            // as a oneliner.
+            r1Head = Math.min(r1Head, r2Head - storage.length) + storage.length;
+            r1Tail = Math.max(r2Tail, r2Tail - storage.length) + storage.length;
+            r2Head = r2Tail = 0; // empty
+        }
+
+        // Reverse base adjustment
+        r1Head += newBase;
+        r1Tail += newBase;
+        r2Head += newBase;
+        r2Tail += newBase;
+
+        if (r1Tail - r1Head >= storage.length || r2Tail - r2Head >= storage.length) {
+            // Evaluate everything
+            normalizeAndEvaluate(0, storage.length, 0, 0);
+        } else {
+            normalizeAndEvaluate(r1Head, r1Tail, r2Head, r2Tail);
+        }
+
+        // Store our computed range
+        calcHead = head;
+        calcTail = tail;
+
+        return treeStorage[1];
+    }
+
+    void normalizeAndEvaluate(final long head1, final long tail1, final long head2, final long tail2) {
```

Comment:
Personally I feel like this method and its various helper methods are too complicated and hard to analyze. I think the job they are trying to accomplish -- namely, keep track of up to three ranges and possibly merge them as you go up levels in the tree -- can be expressed more simply.

I did a prototype implementation in https://github.com/deephaven/deephaven-core/pull/3378 that you may want to take a look at. I took a couple of shortcuts e.g. by calling sort rather than manually ordering my ranges, but overall the resultant code is so short I think you may find the approach valuable.